README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
binaryLogic
===========

[![Build Status](https://travis-ci.org/d4ndo/binaryLogic.png)](https://travis-ci.org/d4ndo/binaryLogic) [![Downloads](https://cranlogs.r-pkg.org/badges/binaryLogic?color=brightgreen)](https://www.r-pkg.org/pkg/binaryLogic) [![CRAN version](https://www.r-pkg.org/badges/version/binaryLogic)](https://cran.r-project.org/package=binaryLogic)

Binary Logic GNU R Package

Convert, negate, shift and rotate binary digits. (switchEndianess, bin2gray, bytesNeeded, binaryPrefix, fillUpToByte).

Installation
------------

devtools are required to install "binaryLogic" from github: [devtools](https://github.com/hadley/devtools)

``` r
library(devtools)

# install 'binaryLogic'
install_github("d4ndo/binaryLogic")
library(binaryLogic)
```

Getting started
---------------

Starting with a simple conversion. Decimal (Base10) to binary (Base2) and vice versa.

``` r
the_answer_to_the_ultimate_question_of_life_the_universe_and_everything <- as.binary(42)

the_answer_to_the_ultimate_question_of_life_the_universe_and_everything
#> [1] 1 0 1 0 1 0

as.numeric(the_answer_to_the_ultimate_question_of_life_the_universe_and_everything)
#> [1] 42

summary(the_answer_to_the_ultimate_question_of_life_the_universe_and_everything)
#>   Signedness  Endianess value<0 Size[bit] Base10
#> 1   unsigned Big-Endian   FALSE         6     42
```

Operator
--------

Behavior »Class Binary«

| Operator                        | Behavior                                                        |
|:--------------------------------|:----------------------------------------------------------------|
| \[== or !=\]                    | Comparision by value.                                           |
| \[&lt;, &lt;= or &gt; , &gt;=\] | Comparision by value.                                           |
| \[+ or -\]                      | Operations by value.                                            |
| \[\*, ^\]                       | Operations by value.                                            |
| \[%/%, %%\]                     | Operations by value                                             |
| \[/\]                           | Not supported.                                                  |
| \[&, ¦, xor\]                   | Bitwise Operations. The smaller vector is filled up with zeros. |
| \[!\]                           | Indicates logical negation (NOT). Bitwise Operations            |

The logical == operator compares every element of the vector (Bitwise comparison). e.g.

``` r
two <- as.binary(2); two <- as.logical(two); two == two;
#> [1] TRUE TRUE
```

The binary == operator compares the value and it does not distinguish between big and little endian.

``` r
two <- as.binary(2); two == two;
#> [1] TRUE
```

BinaryLogic operators:

| Operator                              | Behavior                                                 |
|:--------------------------------------|:---------------------------------------------------------|
| shiftLeft(binary), shiftRight(binary) | shift Operation.                                         |
| rotate(binary)                        | shift Operation.                                         |
| negate(binary)                        | Indicates arithmetic negation. value &lt;- value \* (-1) |
| switchEndianess(binary)               |                                                          |
| bin2gray(binary)                      | convert binary to gray code                              |
| gray2bin(gray)                        | convert gray code to binary                              |

Information
-----------

This class is just not that great at heavy number crunching, but it brings some benefits. Especially if you like to work using vectors in R. The »binary« class inherits from the »logical« class. Some function from package `binaryLogic` can be applied to logical vectors such as shift or rotate (see help).

The internal structure looks like this. It is composed of a »logical vector« and several attributes. In this example(Big-Endian), it corresponds to the value = 2(Base10).

``` r
structure(c(TRUE, FALSE), class = c("binary", "logical"), signed = FALSE, littleEndian = FALSE)
#> [1] 1 0
```

The binary number is represented by a logical vector. The Bit order usually follows the same endianess as the byte order. How to read:

-   Little Endian (LSB) —&gt; (MSB)

-   Big Endian (MSB) &lt;— (LSB)

The Big Endian endianess stores its MSB at the lowest adress. The Little Endian endianess stores its MSB at the highest adress.

e.g.

``` r
b <-binary(8)
b
#> [1] 0 0 0 0 0 0 0 0
```

-   »Little Endian« : MSB at b\[1\] and LSB at b\[8\].

-   »Big Endian« : LSB at b\[1\] and MSB at b\[8\].

##### Signed digit:

Calculation:

The size has to be considerd in a calculation with a signed number. e.G.: An 8 Bit signed number can hold this range of base10 numbers \[127 to -128\]. You will run into a problem if a calculation is outside of this range. The reference is the larger number in size.

Size:

The size »must« be specified. In Byte e.G. (1 Byte = 8 Bit, 2 Byte, .. n Byte). By default it is 2 Byte.

##### Unsigned digit:

Calculation:

An unsigned number will increase it's size when caluculation is not in range.

Size:

The size »can« be specified. In Bit. (1 Bit , 2 Bit, .. n Bit). There is no default size because the size is calculated on the fly.

More Converting
---------------

### Integer

``` r
as.binary(0xAF)
#> [1] 1 0 1 0 1 1 1 1

as.binary(42)
#> [1] 1 0 1 0 1 0

as.binary(42, littleEndian=TRUE)
#> [1] 0 1 0 1 0 1

as.binary(c(0xAF, 0xBF, 0xFF))
#> [[1]]
#> [1] 1 0 1 0 1 1 1 1
#> 
#> [[2]]
#> [1] 1 0 1 1 1 1 1 1
#> 
#> [[3]]
#> [1] 1 1 1 1 1 1 1 1

as.binary(c(2,4,8,16), signed=TRUE, size=1)
#> [[1]]
#> [1] 0 0 0 0 0 0 1 0
#> 
#> [[2]]
#> [1] 0 0 0 0 0 1 0 0
#> 
#> [[3]]
#> [1] 0 0 0 0 1 0 0 0
#> 
#> [[4]]
#> [1] 0 0 0 1 0 0 0 0

as.binary(-1, signed=TRUE, size=1)
#> [1] 1 1 1 1 1 1 1 1
```

other way around

``` r
two <- as.binary(2, signed=TRUE, size=4)
as.integer(negate(two))
#> [1] -2
# or
as.double(two)
#> [1] 2
# alias for
as.numeric(two)
#> [1] 2
```

### Logical

``` r
as.binary(c(1,1,0), signed=TRUE, logic=TRUE)
#> [1] 0 0 0 0 0 1 1 0

as.binary(c(TRUE,TRUE,FALSE), logic=TRUE)
#> [1] 1 1 0

bigEndian <- as.binary(c(1,1,0,0), logic=TRUE)
summary(bigEndian)
#>   Signedness  Endianess value<0 Size[bit] Base10
#> 1   unsigned Big-Endian   FALSE         4     12

littleEndian <- switchEndianess(bigEndian)
print(littleEndian)
#> [1] 0 0 1 1

littleEndian <- as.binary(bigEndian, littleEndian=TRUE)
print(littleEndian)
#> [1] 1 1 0 0

summary(littleEndian)
#>   Signedness     Endianess value<0 Size[bit] Base10
#> 1   unsigned Little-Endian   FALSE         4      3
```

other way around

``` r
as.logical(as.binary(2))
#> [1]  TRUE FALSE
```

### Raw

``` r
b <- as.binary(charToRaw("A"))
summary(b)
#>   Signedness  Endianess value<0 Size[bit] Base10
#> 1   unsigned Big-Endian   FALSE         7     65

as.raw(b)
#> [1] 41
```

### Gray code

``` r
b <- as.binary(0:7, n=3)
g <- lapply(b, bin2gray)
print(g)
#> [[1]]
#> [1] FALSE FALSE FALSE
#> 
#> [[2]]
#> [1] FALSE FALSE  TRUE
#> 
#> [[3]]
#> [1] FALSE  TRUE  TRUE
#> 
#> [[4]]
#> [1] FALSE  TRUE FALSE
#> 
#> [[5]]
#> [1]  TRUE  TRUE FALSE
#> 
#> [[6]]
#> [1] TRUE TRUE TRUE
#> 
#> [[7]]
#> [1]  TRUE FALSE  TRUE
#> 
#> [[8]]
#> [1]  TRUE FALSE FALSE

gray2bin(g[[8]])
#> [1] 1 1 1
```

Special Case
------------

Be aware about this kind of notation »0xAF«. Because Gnu R converts this to an integer first and then it will be converted to a binary digit. This is just a limitation, if you want to use a little endian formation. It can be fixed by using switchEndianess setting the stickyBits=TRUE.

``` r
#Watch out for this notation
as.binary(0xAF, littleEndian=TRUE)
#> [1] 1 1 1 1 0 1 0 1

littleEndian <- switchEndianess(as.binary(0xAF), stickyBits = TRUE)
littleEndian
#> [1] 1 0 1 0 1 1 1 1
summary(littleEndian)
#>   Signedness     Endianess value<0 Size[bit] Base10
#> 1   unsigned Little-Endian   FALSE         8    245
```


NEWS.md
# binaryLogic 0.3.9

* Modified DESCRIPTION to comply with the guidelines of CRAN

# binaryLogic 0.3.8

* fixed travis-ci according https://docs.travis-ci.com/user/languages/r/
* Modified DESCRIPTION to comply with the guidelines of CRAN
* Fixed the url in README.md (https)

# binaryLogic 0.3.7

* minor changes to comply with the guidelines of CRAN.

# binaryLogic 0.3.6

* Optimization of decimal to binary conversion. Remarkable speed gain of 50x. Author A. Rudkin

# binaryLogic 0.3.5

* Fixed the size issue for signed numbers.

# binaryLogic 0.3.4

* Fixed some invalid URL at README.md & README.Rmd

# binaryLogic 0.3.3

* Added a new attribute size n [Bit] to »as.binary« for unsigned number. See `?as.binary()`
* Converter to gray code. See `?bin2gray()` and `?gray2bin()`

# binaryLogic 0.3.2

* Added a new attribute »stickyBits« to switchEndianess. See `?switchEndianess()` for more details.


tests/run-all.R
library(testthat)
library(binaryLogic)


test_package("binaryLogic")

tests/test-BinaryS3.R
library(testthat)
library(binaryLogic)

context("Test rev")

b <- binary(Byte())
sb <- binary(Byte(), signed=TRUE)
lb <- binary(Byte(), littleEndian=TRUE)

test_that("Lost Attributes", {
    expect_that(attr(rev(b), "class"), equals(c("binary","logical")))
    expect_that(attr(rev(b), "signed"), equals(FALSE))
    expect_that(attr(rev(b), "littleEndian"), equals(FALSE))
    expect_that(attr(rev(sb), "signed"), equals(TRUE))
    expect_that(attr(rev(lb), "littleEndian"), equals(TRUE))
})

tests/test-binHelper.R
library(testthat)
library(binaryLogic)

context("binPrefix2Byte")

test_that("Return value", {
    expect_that(binPrefix2Byte(0.5, "KiB"), is_equivalent_to(512))
    expect_that(binPrefix2Byte(1, "KiB"), is_equivalent_to(2^10))
    expect_that(binPrefix2Byte(1, "MiB"), is_equivalent_to(2^20))
    expect_that(binPrefix2Byte(1, "GiB"), is_equivalent_to(2^30))
    expect_that(binPrefix2Byte(1, "TiB"), is_equivalent_to(2^40))
    expect_that(binPrefix2Byte(1, "PiB"), is_equivalent_to(2^50))
    expect_that(binPrefix2Byte(1, "EiB"), is_equivalent_to(2^60))
    expect_that(binPrefix2Byte(1, "ZiB"), is_equivalent_to(2^70))
    expect_that(binPrefix2Byte(1, "YiB"), is_equivalent_to(2^80))
})

#test_that("Warnings", {
#    expect_that(binPrefix2Byte(1,"LiB"), throws_error("Unknown binary prefix"))
#    expect_that(binPrefix2Byte(1.2,"KiB"), shows_message("ceiling called: returns the smallest integer not less than the corresponding element of ret"))
#})

context("Byte")

test_that("Return value", {
    expect_that(Byte(), is_identical_to(8))
})


R/binAdd.R
#' Binary Addition (+)
#' 
#' @description Adds two binary numbers. (x + y)
#' @details Little-Endian and unsigned is not supported at the moment. No floating point supported.
#' if x or y is signed the return value will also be signed.
#' @usage binAdd(x, y)
#' @param x summand 1 (binary vector)
#' @param y summand 2 (binary vector)
#' @return The sum of x and y. Returns a binary vector.
#' @examples
#' five <- as.binary(5); ten <- as.binary(10);
#' as.numeric(binAdd(ten, five))
#' binAdd(as.binary(c(0,1)),as.binary(c(1,0)))
#' @seealso base::as.logical , base::is.logical, base::raw
#' @export
binAdd <- function(x, y) {
    #Should be implemented in C (internal generic). Very slow.
    if (missing(x)) stop("x is missing.")
    if (missing(y)) stop("y is missing.")
    stopifnot(is.binary(x))
    stopifnot(is.binary(y))
    
    x_signed <- attributes(x)$signed
    x_littleEndian <- attributes(x)$littleEndian
    y_signed <- attributes(y)$signed
    y_littleEndian <- attributes(y)$littleEndian
    
    l <- list(class=c("binary","logical"),
              signed=FALSE,
              littleEndian=x_littleEndian)
    
    if (x_signed || y_signed) l$signed <- TRUE
    if (x_littleEndian) y <- switchEndianess(y)
    if (y_littleEndian) x <- switchEndianess(x)    
    
    if (length(x) >= length(y))
    {
        MAX <- length(x)
        a <- rep(FALSE,length(x)-length(y))
        y <- c(a,y)
    } else {
        MAX <- length(y)
        a <- rep(FALSE,length(y)-length(x))
        x <- c(a,x)
    }
    
    ret = binary(MAX)
    temp = binary(MAX+1)
    ret[MAX] <- xor(x[MAX],y[MAX])
    if ((isTRUE(as.logical(x[MAX])) && isTRUE(as.logical(y[MAX])))) temp[MAX+1] <- TRUE
    
    if (MAX > 2)
    {
        for(i in (MAX-1):1)
        {
            ret[i] <- xor(x[i],y[i])
            ret[i] <- xor(ret[i],temp[i+2])
            if(((isTRUE(as.logical(x[i])) && isTRUE(as.logical(y[i]))) || 
                   (isTRUE(as.logical(x[i])) && isTRUE(as.logical(temp[i+2]))) ||
                   (isTRUE(as.logical(y[i])) && isTRUE(as.logical(temp[i+2]))))) temp[i+1] <- TRUE
        }
    }
    if (temp[2] && !l$signed) {
        ret <- c(T,ret)
        #l$signed <- TRUE
        #l$littleEndian <- FALSE
    }
    # reverse big endian.
    return(loadAttributes(ret, l))
}

operator.R
Operator <- c("[== or !=]", "[&lt;, &lt;= or > , >=]", "[+ or -]", "[*, ^]", "[%/%, %%]","[/]", "[&, &#166;, xor]", "[!]")
Behavior <- c("Comparision by value.", "Comparision by value.", "Operations by value.", "Operations by value.", "Operations by value", " Not supported.", "Bitwise Operations. The smaller vector is filled up  with zeros.", "Indicates logical negation (NOT). Bitwise Operations")
op <- as.data.frame(cbind(Operator,Behavior))

Operator <- c("shiftLeft(binary), shiftRight(binary)", "rotate(binary)", "negate(binary)", "switchEndianess(binary)", "bin2gray(binary)", "gray2bin(gray)")
Behavior <- c("shift Operation.", "shift Operation.", "Indicates arithmetic negation. value <- value * (-1)", " ", "convert binary to gray code", "convert gray code to binary")
op2 <- as.data.frame(cbind(Operator,Behavior))

R/binaryLogic.R
#' Binary Logic GNU R Package

#' @description 
#' This package contains the \strong{\link{binary}} S3 class.
#' A data object can be instantiated to store a binary number(Base2).
#'
#' It can be used to convert, negate, shift or rotate the binary number.
#' (switchEndianess, BytesNeeded, binaryPrefix, fillUpToByte).
#' 
#' Binary operators:
#' \itemize{
#'  \item \strong{==} , \strong{!=} , \strong{<} , \strong{<=} , \strong{>} , \strong{>=}
#'  \item \strong{+} , \strong{-} , \strong{^}, \strong{*}
#'  \item \strong{&} , \strong{|} , \strong{xor} (Logical Operator. Bitwise operation. The smaller vector is added up with zeros)
#'  \item \strong{!} (Indicates logical negation (NOT). Bitwise Operations)
#' }
#' 
#' binaryLogic functions:
#' \itemize{
#'  \item \link{shiftLeft}(binary) , \link{shiftRight}(binary)
#'  \item \link{rotate}(binary)
#'  \item \link{negate}(binary)
#'  \item \link{switchEndianess}(binary)
#' }
#'
#' Additional function:
#' \itemize{
#'  \item \link{fillUpToByte}, \link{fillUpToBit}
#'  \item \link{bytesNeeded}
#'  \item \link{binPrefix2Byte}
#'  \item  \link{Byte}
#' }
#'
#' @details
#' This \strong{\link{binary}} class is just not that great at heavy number crunching, but it brings some benefits.
#' Especially if you like to work using vectors in R. It inherits from the \emph{logical} class.
#' Some function from package \pkg{binaryLogic} can be applied to \emph{logical} vectors. Such as shift or rotate (see help).
#' 
#' The internal structure looks like this:
#' 
#' \code{structure(c(TRUE, FALSE), class = c("binary", "logical"), signed = FALSE, littleEndian = FALSE)}
#' 
#' It is composed of a \emph{logical} vector and several attributes. This structure shows a big endian number, 
#' it corresponds to the value = 2(Base10).
#' @docType package
#' @name binaryLogic
NULL

R/BinaryS3.R
#' Binary digit.
#' 
#' @description Create objects of type binary.
#' @details The binary number is represented by a \emph{logical} vector.
#' The bit order usually follows the same endianess as the byte order.
#' How to read:
#' \itemize{
#' \item Little Endian    (LSB) ---> (MSB)
#' \item Big Endian       (MSB) <--- (LSB)
#' }
#' The Big Endian endianess stores its MSB at the lowest adress. 
#' The Little Endian endianess stores its MSB at the highest adress.
#' 
#' e.g. b <-binary(8).
#' \itemize{
#' \item "Little Endian" : MSB at b[1] and LSB at b[8].
#' \item "Big Endian" : LSB at b[1] and MSB at b[8].
#' }
#' No floating-point support.
#' @usage binary(n, signed=FALSE, littleEndian=FALSE)
#' @param n length of vector. Number of bits
#' @param signed  TRUE or FALSE. Unsigned by default. (two's complement)
#' @param littleEndian if TRUE. Big Endian if FALSE.
#' @return a vector of class binary of length n. By default filled with zeros(0).
#' @examples
#' b <- binary(8)
#' summary(b)
#' b <- binary(16, signed=TRUE)
#' summary(b)
#' b <- binary(32, littleEndian=TRUE)
#' summary(b)
#' @seealso \link{as.binary} and \link{is.binary}.
#' @export
binary <- function(n, signed=FALSE, littleEndian=FALSE) {
    x <- logical(n)
    class(x) <- c("binary", "logical")
    attr(x, "signed") <- signed
    attr(x, "littleEndian") <- littleEndian
    if (signed) { x <- fillUpToByte(x) }
    return(x)
}

#' as binary digit.
#' 
#' @description Converts an integer (Base10) 
#' to a binary (Base2) number. It also converts a logical vector 
#' to a binary (Base2) number (see examples).
#' @details The binary number is represented by a logical vector.
#' The bit order usually follows the same endianess as the byte order.
#' No floating-point support. If logic is set to TRUE an integer vector 
#' is intepreted as a logical vector (>0 becomes TRUE and 0 becomes FALSE)
#' \itemize{
#' \item Little Endian    (LSB) ---> (MSB)
#' \item Big Endian       (MSB) <--- (LSB)
#' }
#' Auto switch to signed if num < 0.
#' @usage as.binary(x, signed=FALSE, littleEndian=FALSE, size=2, n=0, logic=FALSE)
#' @param x integer or logical vector.
#' @param signed  TRUE or FALSE. Unsigned by default. (two's complement)
#' @param littleEndian if TRUE. Big Endian if FALSE.
#' @param size in Byte. Needed if \bold{signed} is set. (by default 2 Byte)
#' @param n in Bit. Can be set if \bold{unsigned} is set to TRUE. (by default 0 Bit = auto)
#' @param logic If set to TRUE, x is expected as logical vector.
#' @return a vector of class binary.
#' @examples
#' as.binary(0xAF)
#' as.binary(42)
#' as.binary(42, littleEndian=TRUE)
#' as.binary(c(0xAF, 0xBF, 0xFF))
#' as.binary(c(2,4,8,16,32), signed=TRUE, size=1)
#' as.binary(-1, signed=TRUE, size=1)
#' as.binary(1:7, n=3)
#' as.binary(sample(2^8,3),n=8)
#' as.binary(c(1,1,0), signed=TRUE, logic=TRUE)
#' as.binary(c(TRUE,TRUE,FALSE), logic=TRUE)
#' @seealso \link{is.binary} and \link{binary}
#' @export
as.binary <- function(x, signed=FALSE, littleEndian=FALSE, size=2, n=0, logic=FALSE) {
    if (!inherits(x, "binary")) {
        if (logic == FALSE)
        {
            x <- binSeq(x, signed=signed, littleEndian=littleEndian, size=size, n=n)
        } else {
            x <- as.logical(x)
            class(x) <- c("binary", "logical")
            attr(x, "signed") <- signed
            attr(x, "littleEndian") <- littleEndian
            if (signed) x <- fillUpToByte(x)
            else x <- fillUpToBit(x, n)
        }
    } else {
        l <- saveAttributes(x)
        if (signed) {
            l$signed <- TRUE
            #attr(x, "signed") <- TRUE
            x <- fillUpToByte(x)
        } else {
            l$signed <- FALSE
            x <- fillUpToBit(x, n)
            #attr(x, "signed") <- FALSE
        }
        if (littleEndian) {
            l$littleEndian <- TRUE
        } else {
            l$littleEndian <- FALSE
        }
        x <- loadAttributes(x, l)
    }
    return(x)
}

#' is Binary Vector
#' 
#' @description test for object "binary".
#' @usage is.binary(x)
#' @param x object to test.
#' @return TRUE or FALSE.
#' @seealso \link{as.binary} and \link{binary}
#' @export
is.binary <- function(x) {
    return(inherits(x, "binary"))
}

#' Print method for binary number.
#' 
#' @description This method prints the binary number.
#' @param x any binary number.
#' @param ... further arguments.
#' @return Output in ones and zeros (binary vector).
#' @seealso \link{summary.binary} provides some additional information.
#' @method print binary
#' @export
print.binary <- function(x, ...) {
    x <- ifelse(x, as.integer(1), as.integer(0))
    attributes(x) <- NULL
    NextMethod(print.default)
}

#' Summary method for binary number.
#' 
#' @description This method provides information about the attributes of the binary number.
#' @param object binary number.
#' @param ... further arguments.
#' @return Contains the following information:
#' \itemize{
#' \item Signedness : unsigned or signed
#' \item Endianess : Big-Endian or Little-Endian
#' \item value<0 : negative or positve number
#' \item Size[bit] : Size in bit
#' \item Base10 : Decimal(Base10) number.
#' }
#' @seealso \link{print.binary}
#' @method summary binary
#' @export
summary.binary <- function(object, ...) {
    #I'm not sure if this is the way to do it. just printing a dataframe.
    l <- saveAttributes(object)

    tableHead <- character(3)
    signedness <- character(1)
    endianess <- character(1)
    size <- length(object)
    neg <- logical(1)
    
    tableHead <- c("Signedness", "Endianess", "value<0", "Size[bit]", "Base10")
    if (l$signed) {
        signedness <- "signed"
        if (l$littleEndian) {
            if (object[length(object)]) neg <- TRUE else neg <- FALSE
        } else {
            if (object[1]) neg <- TRUE else neg <- FALSE
        }
    } else {
        signedness <- "unsigned"
        neg <- FALSE
    }
    if (l$littleEndian) { endianess <- "Little-Endian" } else { endianess <- "Big-Endian" }
    df <- data.frame(signedness, endianess, neg, size, bin2dec(object))
    colnames(df) <- tableHead
    print(df)
}

#' @export
as.raw.binary <- function(x) {
    #b <- as.binary(rawToBits(r)) from raw to binary
    l <- saveAttributes(x)
    x <- fillUpToByte(x, size=bytesNeeded(length(x)))
    xx <- logical(0)
    
    if (!l$littleEndian) x <- rev(x)
    if (l$littleEndian) {
        x <- as.logical(x)
        dim(x) <- c(4, (2 * bytesNeeded(length(x))))
        for(i in seq(1, (2*bytesNeeded(length(x)) - 1), by = 2)) xx <- c(xx, c(x[,i+1], x[,i]))
        x <- xx
    }    
    x <- packBits(x)
    if(!l$littleEndian) x <- rev(x)
    NextMethod(.Generic)
} #rseq = function(x,to=1) NROW(x):to

##' @export
#as.hexmode.binary <- function(x) {
#    x <- bin2dec(x)
#    NextMethod(.Generic)
#}

##' @export
#as.octmode.binary <- function(x) {
#    x <- bin2dec(x)
#    NextMethod(.Generic)
#}

#' @export
as.character.binary <- function(x, ...) {
    x <- as.integer(as.logical(x))
    NextMethod(.Generic, ...)
}

#' @export
as.integer.binary <- function(x, ...) {
    #test for .Machine$integer.max
    x <- bin2dec(x)
    NextMethod(.Generic, ...)
}

#' @export
as.double.binary <- function(x, ...) {    
    x <- bin2dec(x)
    NextMethod(.Generic, ...)
}

#' Group Generic Ops
#' 
#' Group generic Ops operators
#' @param e1 e1
#' @param e2 e2
#' @method Ops binary
#' @export
Ops.binary <- function(e1, e2) {
    # Group Generic "Ops"
    boolean <- switch(.Generic,  '/' = TRUE, FALSE)
    if (boolean) {
        stop("Division is not supported. For the time being")
    }
    boolean <- switch(.Generic,  '+' =, '-' = TRUE, FALSE)
    if (boolean) {
        l1 <- saveAttributes(e1)
        ret <- NextMethod(.Generic)
        x <- loadAttributes(ret,l1)
        return(x)
    }
    boolean <- switch(.Generic,  '*' =, '^' =, '%%' =, '%/%' = TRUE, FALSE)
    if (boolean) {
        l1 <- saveAttributes(e1)
        l2 <- saveAttributes(e2)
        le1 <- length(e1)
        le2 <- length(e2)
        e1 <- as.numeric(e1)
        e2 <- as.numeric(e2)
        if (l1$signed || l2$signed)
        {
            ret <- NextMethod(.Generic)
            ret <- as.binary(ret, signed=TRUE, size=(greater(le1, le2)/byte()))
            x <- loadAttributes(ret, l1)
        } else {
            ret <- NextMethod(.Generic)
            ret <- as.binary(ret)
            x <- loadAttributes(ret, l1)
        }
        return(x)
    }
    boolean <- switch(.Generic,  '&' =, '|' =, 'xor' = TRUE, FALSE)
    if (boolean) {
        l1 <- saveAttributes(e1)
        if (length(e1) < length(e2))
        {
            e1 <- fillUpToBit(e1, value=FALSE, n=length(e2))
        } else if (length(e2) < length(e1)) {
            e2 <- fillUpToBit(e2, value=FALSE, n=length(e1))
        }
        ret <- NextMethod(.Generic)
        x <- loadAttributes(ret, l1)
        return(x)
    }
    boolean <- switch(.Generic,  '!' = TRUE, FALSE)
    if (boolean) {
        l1 <- saveAttributes(e1)
        ret <- NextMethod(.Generic)
        x <- loadAttributes(ret, l1)
        return(x)
    }
    boolean <- switch(.Generic,  '<' =, '<=' =, '>' =, '>=' = TRUE, FALSE)
    if (boolean) {
        e1 <- as.numeric(e1)
        e2 <- as.numeric(e2)
        ret <- NextMethod(.Generic)
        return(ret)
    }
}

#' @export
'+.binary' <- function(x, y) {
    binAdd(x, y)
}

#' @export
'-.binary' <- function(x, y) {
    binAdd(x, negate(y))
}

#' @export
'==.binary' <- function(x, y) {
    # attributes are saved @ group generic Ops.  
    if (attributes(x)$littleEndian) x <- switchEndianess(x)
    if (attributes(y)$littleEndian) y <- switchEndianess(y)
    
    if (length(x) > length(y)) {
        delta <- bytesNeeded(length(x))
        if (attributes(y)$signed) {
            y <- fillUpToByte(y, value=TRUE, size=delta)
        } else {
            y <- fillUpToByte(y, value=FALSE, size=delta)
        }
    } else if (length(x) < length(y)) {
        delta <- bytesNeeded(length(y))
        if (attributes(x)$signed) {
            x <- fillUpToByte(x, value=TRUE, size=delta)
        } else {
            x <- fillUpToByte(x, value=FALSE, size=delta)
        }
    }
    return(all(!xor(x, y)))
}

#' @export
'!=.binary' <- function(x, y) {
    # attributes are saved @ group generic Ops.
    !(x == y)
}

#' @export
'[.binary' <- function(x, i, j, drop=TRUE) {
    # this should not become a group generic. This function is an internal generic.
    l <- saveAttributes(x)
    
    ret <- NextMethod(.Generic, drop=drop)
    
    x <- loadAttributes(ret,l)
    return(x)
}

#' @export
rev.binary <- function(x) {
    # this should not become a group generic. This function is an internal generic.
    l <- saveAttributes(x)
    #x = x[rseq(x)] rseq = function(x,to=1) NROW(x):to
    x <- x[length(x):1]
    #ret <- NextMethod(.Generic)

    x <- loadAttributes(x,l)
    return(x)
}

#' saveAttributes
#'  
#' Helper function save Attributes
#' @usage saveAttributes(x)
#' @param x x
#' @export
saveAttributes <- function(x) {
    if(is.binary(x)) l <- list(class=c("binary","logical"),
                                     signed=attr(x,"signed"),
                                     littleEndian=attr(x,"littleEndian"))
    else l <- list(class=class(x))
    return(l)
}

#' loadAttributes
#'  
#' Helper function load Attributes
#' @usage loadAttributes(x, l)
#' @param x x
#' @param l l 
#' @export
loadAttributes <- function(x, l) {
    class(x) <- l$class
    attr(x, "signed") <- l$signed
    attr(x, "littleEndian") <- l$littleEndian
    return(x)
}

# Helper function
dec2bin <- function(num, signed=FALSE, littleEndian=FALSE, size=2, n=0) {
    # Very slow with negative numbers. (maybe binAdd)
    if (signed && (((num > ((2^(size*byte())/2)-1))) || 
                       (num < ((-1)*(2^(size*byte())/2)))))
        stop("Out of Range. Please increase the size[Byte]")
    
    if (signed == FALSE && n > 0 && 
        (((num > ((2^(n))-1))) || (num < (0))))
        stop("Out of Range. Please increase the size n [Bit], or set it to 0=auto")

    l <- list(class=c("binary","logical"),
              signed=signed,
              littleEndian=littleEndian)
    
    #needed for hex input
    num <- as.numeric(num)
    
    neg = FALSE
    if (num < 0) {
        l$signed <- TRUE
        neg = TRUE
        num = abs(num)
    }
    
    #global read only for function h(num)
    ret <- h(num)
    
    if(l$signed) {
        b <- logical(size*byte()) 
        } else {
        if (n > 0) { 
            b <- logical(n)
        } else {
            b <- logical(max(ret)-1)
        }
    }
    #do some optimization here.
    for (i in seq(length(b))) b[i] <- any(ifelse((ret-1)==i, TRUE, FALSE))
    
    if (neg)
    {
        b <- !rev(b)
        b <- rev(binAdd(as.binary(b, signed=TRUE, logic=TRUE), as.binary(1, signed=TRUE, logic=TRUE)))
    }
    if(!l$littleEndian) b <- rev(b)
    loadAttributes(b, l)
    return(loadAttributes(b, l))
}

# Helper function to convert decimal to binary
h <- function(x) {
    # Basic error handling: function expects single positive integer
    if(x %% 1 != 0) { stop("Non-integer input to h()") }
    if(x < 0) { stop("Negative input to h()") }

    # Special case
    if(x == 0) { return(c(1, 0)) }

    # Only generate as many comparison digits as we need
    num_digits = floor(log2(x)) + 1

    # Get x %% each power of two; points in the vector where this
    # changes are active. Because my function has bit order reversed
    # versus the original h() function, we need to go num_digits -
    # the active bits. The +2 and append 0 are to match the original
    # h behavior
    c(num_digits + 2 - which(diff(x %% 2^(num_digits:0)) != 0), 0)
}

# Helper function
bin2dec <- function(bin) {
    #could be implemented in C. But it is not that slow.
    signed <- attributes(bin)$signed
    littleEndian <- attributes(bin)$littleEndian
    
    if(!littleEndian) { bin <- rev(bin) }
    
    bin <- as.integer(as.logical(bin))
    i = length(bin)-1
    numeric = 0
    first <- TRUE
    
    for(d in rev(bin))
    {
        if ((signed) & (first)) {
            numeric <- (-1 * d * (2^i))
            i <- i - 1
            first <- FALSE
        } else {
            numeric = (numeric + d * (2^i))
            i <- i - 1
        }
    }
    return(numeric)
}

# Helper function
greater <- function(e1, e2) {
    if (e1 > e2)
    {
        return(e1)
    } else {
        return(e2)
    }
}
